File: src/core/scala/mrtjp/core/fx/GroupAction.scala
/*
 * Copyright (c) 2015.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.core.fx

import mrtjp.core.fx.particles.CoreParticle

import scala.collection.mutable.{Seq => MSeq}

class GroupAction extends ParticleAction {
    var actions = MSeq[ParticleAction]()

    override def tickLife() {
        super.tickLife()
        actions.foreach(_.tickLife())
    }

    override def runOn(p: CoreParticle, frame: Float) {
        super.runOn(p, frame)

        actions.foreach { a =>
            if (!a.isFinished) {
                a.runOn(p, frame)
            }
        }

        if (actions.forall(_.isFinished)) {
            isFinished = true
        }
    }

    override def operate(p: CoreParticle, time: Double) {}

    override def compile(p: CoreParticle) {
        super.compile(p)
        actions.foreach(_.compile(p))
    }

    override def reset() {
        super.reset()
        actions.foreach(_.reset())
    }

    override def copy = ParticleAction.group(actions.map(_.copy).toList: _*)
}

File: core/src/main/java/mrtjp/projectred/core/client/particle/GroupAction.java
package mrtjp.projectred.core.client.particle;

public class GroupAction extends ParticleAction {

    private final ParticleAction[] actions;

    public GroupAction(ParticleAction... actions) {
        this.actions = actions;
    }

    @Override
    public void tick() {
        super.tick();
        for (ParticleAction action : actions) {
            action.tick();
        }
    }

    @Override
    public void operate(BaseActionParticle particle, float partialTick) {
        super.operate(particle, partialTick);
        boolean allFinished = true;
        for (ParticleAction action : actions) {
            action.operate(particle, partialTick);
            if (!action.isFinished()) {
                allFinished = false;
            }
        }
        finished = allFinished;
    }

    @Override
    public ParticleAction copy() {
        ParticleAction[] newActions = new ParticleAction[actions.length];
        for (int i = 0; i < actions.length; i++) {
            newActions[i] = actions[i].copy();
        }
        return new GroupAction(newActions);
    }

    @Override
    public void beginAction(BaseActionParticle particle) {
        for (ParticleAction action : actions) {
            action.beginAction(particle);
        }
    }

    @Override
    public void operateAction(BaseActionParticle particle, double time) {
        // Action does nothing
    }
}
